<a href="https://colab.research.google.com/github/gabyx7677/Proyectos/blob/Proyectos-Triple-Ten/Proyecto_Sprint_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Déjame escuchar música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa de ellos y tomar decisiones en consecuencia. Esto implica pasar por diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investiguemos, necesitamos formular hipótesis que después podamos probar. A veces, aceptaremos estas hipótesis; otras veces, las rechazaremos. Para tomar las decisiones adecuadas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de música online para probar la hipótesis que planteamos a continuación y compararás el comportamiento de los usuarios de estas dos ciudades.

### Objetivo:
Probar la hipótesis:
1. La actividad de los usuarios difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento de los usuarios se almacenan en el archivo `/datasets/music_project_en.csv`. No hay información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, debes evaluar la calidad de los datos y ver si los problemas son significativos. Más tarde, durante el preprocesamiento de datos, deberás abordar los problemas más críticos.

Tu proyecto contará con estas tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de los datos.
 3. Prueba de la hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importa pandas
import pandas as pd

Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Lee el archivo y almacénalo en df
df = pd.read_csv("/content/music_project_en.csv")

Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtén las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [4]:
# Obtén la información general sobre nuestros datos
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí` (no uso de snake_case).




### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1. Las filas representan reproducciones individuales de canciones. Cada columna almacena atributos específicos: por ejemplo, 'userID' identifica al usuario, 'Track' el título, 'time' la hora, etc. Podemos entender esto por los nombres de las columnas y su descripción. Además, inspeccionar con df.info() y df.head() ayuda a confirmar qué tipo de datos contiene cada una. Las series contienen datos categoricos que se pueden analizar de diferentes maneras.

2. Depende de la hipótesis, pero en general, con 'userID', 'genre', 'time', 'Day' y 'City', hay suficiente información para analizar patrones de reproducción por usuario, género, hora y ubicación. Si la hipótesis fuera más compleja (como ingresos, edad o suscripción), sí faltarían datos.

3. Valores ausentes en 'Track' o 'artits', lo cual puede distorsionar análisis, se repiten filas idénticas de reproducción Tipos incorrectos, como 'time' siendo texto en lugar de tipo datetime lo cual cierra la posibilidad a poder manipular ese dato en horas, minutos y segundos.

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Muestra los nombres de las columnas
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case.


Anteriormente, aprendiste una forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora.

Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
# Bucle en los encabezados que lo pone todo en minúsculas
for col in df.columns:
    df.columns = df.columns.str.lower()

df.columns

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')

Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [7]:
# Bucle en los encabezados que elimina los espacios
for col in df.columns:
    df.columns = df.columns.str.strip()

df.columns

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
# Cambia el nombre de la columna "userid"
df.rename(columns={"userid" : "user_id"}, inplace=True)
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
# Comprueba el resultado: lista de encabezados

df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [10]:
# Calcula el número de valores ausentes

df_ausentes1 = df.isnull().sum()
df_ausentes2 = df.isna().sum()
print(df_ausentes1)
print()
print(df_ausentes2)

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan de la misma forma a la investigación. Por ejemplo, los valores ausentes en 'track' y 'artist' no son cruciales para el análisis, ya que estos datos son más descriptivos que analíticos. Por eso, puedes reemplazarlos directamente con un valor predeterminado como el string 'unknown' (desconocido).

En cambio, los valores ausentes en 'genre' sí pueden influir en la comparación entre las preferencias musicales de Springfield y Shelbyville. En un escenario real, lo ideal sería investigar por qué faltan estos datos e intentar recuperarlos. Pero en este proyecto no tenemos esa posibilidad, así que deberás rellenar esos valores con un valor predeterminado.


Como vimos anteriormente en las lecciones, la mejor manera de hacerlo es crear una lista con los nombres de las columnas que necesitan reemplazarse.  Luego, utilizar esta lista para iterar sobre cada columna y realizar el reemplazo correspondiente.

Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [11]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'

for col in ['track', 'artist', 'genre']:
    df[col].fillna('unknown', inplace=True)

print(df.sample(10))

        user_id                                              track  \
511    C8AF03B0                                                Hey   
46518  EA721BBE                                             Refuge   
10793  932B0FBE  Still Have a Dream (Official Street Parade 200...   
4585   38303C5B                                      Magical Night   
22599  7E39A5DF                     Can't We Talk Over? (02- ?-40)   
34615  E4043051                                      Favorite Song   
52977  A123316B                                     Smoke / Dreams   
4563   E951E691                                      Última Jugada   
62113  68AF20CA                                   Secret Agent Man   
50796  8B9199FE                                       Ghostbusters   

                                 artist       genre         city      time  \
511                                 Kel      ruspop  Shelbyville  20:27:28   
46518           While the Heart Becomes       metal  Springfield  21:15:2

<ipython-input-11-596ad95ef0d0>:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna('unknown', inplace=True)


Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [12]:
# Cuenta los valores ausentes

df_ausentes1 = df.isnull().sum()
df_ausentes2 = df.isna().sum()
print(df_ausentes1)
print()
print(df_ausentes2)

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [13]:
# Cuenta los duplicados explícitos

print(df.duplicated().sum())

3826


Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
# Elimina los duplicados explícitos
df.drop_duplicates(inplace=True)

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
# Comprueba de nuevo si hay duplicados

print(df.duplicated().sum())

0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [16]:
# Inspecciona los nombres de géneros únicos
sorted(df["genre"].unique())


['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',
 'ethnic',
 'eurofolk',
 'european',
 'expe

Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Algunos de los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_genres()` que tendrá dos parámetros:

* `wrong_genres`: una lista con todos los valores que deben ser reemplazados.  
* `correct_genre`: un string que se utilizará como valor de reemplazo.

El objetivo de esta función es **corregir los valores en la columna `'genre'` del DataFrame `df`**, reemplazando cada valor de la lista `wrong_genres` por `correct_genre`.

Etapa 2.12.
1. Define una función llamada `replace_wrong_genres()` que reciba dos parámetros: `wrong_genres` y `correct_genre`.

2. Dentro de la función, utiliza un bucle `for` para iterar sobre cada valor en la lista `wrong_genres`.

3. En cada iteración, accede a la columna `'genre'` del DataFrame `df` y utiliza el método `.replace()` para sustituir el valor incorrecto por `correct_genre`.

4. Llama a la función y pasa como argumentos:
   - Una lista con los duplicados implícitos: `['hip', 'hop', 'hip-hop']`
   - El string de reemplazo: `'hiphop'`

In [17]:
from os import replace
# Función para reemplazar los duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)

Etapa 2.13. Ahora, llama a `replace_wrong_genres()` y pásale estos argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los sustituya por `hiphop`:

In [18]:
# Elimina los duplicados implícitos
replace_wrong_genres(['hip', 'hop', 'hip-hop'], 'hiphop')

Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
# Comprueba de nuevo los duplicados implícitos
sorted(df["genre"].unique())



['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',
 'ethnic',
 'eurofolk',
 'european',
 'expe

[Volver a Contenidos](#back)

### Observaciones <a id='data_preprocessing_conclusions'></a>

`Redacta un breve resumen de lo que descubriste al analizar los datos. Tu respuesta debe identificar los problemas detectados, explicar cómo los resolviste y describir cómo esas acciones mejoran la calidad del análisis.`

Etapa 2.14. Descríbelo aquí.
 #----->>> REVISAR
  
Resumen del análisis y limpieza de datos

Durante el proceso de análisis inicial, se detectaron varios problemas que podrían afectar la calidad y precisión del análisis posterior. Entre ellos, se identificaron:

Duplicados y valores ausentes, que pueden distorsionar resultados y sesgar interpretaciones.

Inconsistencias en las categorías, con valores repetidos por ambigüedad o diferencias en mayúsculas/minúsculas.

Espacios en blanco al inicio o final en nombres de columnas, que afectan la correcta manipulación y filtrado.

Nombres de columnas en formatos poco estandarizados, dificultando la legibilidad y el manejo automatizado.

Para resolver estas cuestiones, se aplicaron las siguientes acciones:

Se convirtieron todos los encabezados a minúsculas (lower case) para homogeneizar categorías y evitar duplicados invisibles.

Se eliminaron filas duplicadas y se rellenaron valores ausentes de manera adecuada para mantener la integridad del conjunto de datos.

Se verificó la unicidad y coherencia de las categorías únicas para evitar ambigüedades.

Se eliminaron espacios en blanco superfluos en columnas y datos para asegurar la consistencia en los filtros y comparaciones.

Se renombraron las columnas al formato snake_case para mejorar la claridad, uniformidad y facilitar la escritura del código.

Estas acciones mejoran significativamente la calidad del análisis al asegurar que los datos estén limpios, homogéneos y estructurados de manera óptima, reduciendo errores y facilitando la interpretación y generación de insights confiables.

[Volver a Contenidos](#back)

## Etapa 3. Prueba de la hipótesis <a id='hypothesis'></a>

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

La hipótesis que queremos probar plantea que existen diferencias en la forma en que los usuarios de Springfield y Shelbyville consumen música.
Para analizar esto, nos enfocaremos en los datos correspondientes a tres días específicos de la semana: lunes, miércoles y viernes.

El análisis consistirá en comparar la cantidad de canciones reproducidas por los usuarios de cada ciudad en esos días. Esto nos permitirá observar posibles patrones o diferencias en los hábitos de consumo entre Springfield y Shelbyville.

Para llevar a cabo este análisis, es importante seguir el enfoque de dividir-aplicar-combinar, del que ya hablamos en la lección. En este caso:

*  Dividir: separamos los datos en grupos según la ciudad.

*  Aplicar: dentro de cada grupo, contamos cuántas canciones se reprodujeron.

*  Combinar: reunimos los resultados en una estructura que nos permita comparar fácilmente ambas ciudades.

Este procedimiento debe repetirse de forma independiente para cada uno de los tres días seleccionados. El resultado final debe mostrar el número de reproducciones por ciudad en cada uno de esos días.

Una posible forma de estructurar el código sería con la siguiente expresión:

`df.groupby(by='...')['...'].method()`

Deberás completar los argumentos correspondientes para agrupar por ciudad y contar las canciones reproducidas. Este enfoque te dará una visión clara y comparativa del comportamiento de los usuarios en ambas ciudades.

Etapa 3.1. Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna 'track' como referencia.

In [20]:
# Cuenta las canciones reproducidas en cada ciudad usando la división, aplicación y combinación

shelbyville = df[df["city"] == "Shelbyville"]["track"].count()
springfield = df[df["city"] == "Springfield"]["track"].count()

print(f" Shelbyville: {shelbyville} \n Springfield: {springfield}")

# Cuenta las canciones reproducidas en cada ciudad usando la función groupby
df.groupby("city")["track"].count()

 Shelbyville: 18512 
 Springfield: 42741


city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

`Redacta brevemente tus observaciones sobre los resultados. ¿Qué diferencias notaste entre Springfield y Shelbyville? ¿A qué podrían deberse esas diferencias?`

Etapa 3.2. Escribe tus observaciones aquí.

Los datos muestran una diferencia notable en la cantidad de reproducciones entre Springfield (42,741) y Shelbyville (18,512). Esta diferencia podría explicarse por varios factores. En primer lugar, Springfield podría tener una población mayor, lo que naturalmente implicaría un mayor volumen de usuarios y, por ende, más reproducciones. También podría influir el perfil demográfico o cultural de cada ciudad; por ejemplo, Springfield podría contar con una población más joven o con mayor acceso a tecnología y plataformas de música digital. Finalmente, también puede haber factores locales como eventos, campañas publicitarias o presencia de artistas locales que hayan impulsado la actividad musical en esa ciudad.

Etapa 3.3.
1. Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes, miércoles y viernes.

2. Utiliza el mismo método de conteo que antes, pero ahora cambia la columna de agrupación para enfocarte en el día.

3. Esto te permitirá identificar posibles patrones de escucha según el día de la semana.


In [21]:
# Calcula las canciones reproducidas en cada uno de los tres días usando la división, aplicación y combinación


lunes = df[df["day"] == "Friday"]["day"].count()
miercoles = df[df["day"] == "Monday"]["day"].count()
viernes = df[df["day"] == "Wednesday"]["day"].count()

print(f" lunes: {lunes} \n miercoles: {miercoles} \n viernes: {viernes}")

# Calcula las canciones reproducidas en cada uno de los tres días usando  la función groupby
df.groupby("day")["day"].count()

 lunes: 21840 
 miercoles: 21354 
 viernes: 18059


day
Friday       21840
Monday       21354
Wednesday    18059
Name: day, dtype: int64

`Describe brevemente qué observaste al comparar los lunes, miércoles y viernes. ¿Hubo algún día con menos actividad? ¿Cambian las conclusiones si analizas cada ciudad por separado?`

Etapa 3.4. Escribe tus observaciones aquí.

Al comparar los datos por día, se observa que el viernes registró la mayor cantidad de reproducciones (21,840), seguido de cerca por el lunes (21,354), mientras que el miércoles mostró una disminución notable (18,059). Esto sugiere un patrón donde la actividad musical tiende a ser más alta a inicios y fines de la semana laboral, posiblemente por cambios en los hábitos de las personas: el lunes puede representar un inicio de semana acompañado de música para motivarse, mientras que el viernes refleja un ambiente más relajado o festivo. En contraste, el miércoles, como día medio, podría asociarse con menor tiempo libre o menos disposición para consumir música.

Si analizamos los datos por ciudad, las conclusiones podrían variar. Por ejemplo, una ciudad con mayor población estudiantil o con fuerte vida nocturna podría mostrar picos diferentes, por lo que sería útil desglosar por ciudad para identificar patrones locales más específicos.

Hasta ahora has aprendido a contar entradas agrupándolas por un solo criterio, como la ciudad o el día de la semana. Ahora vamos a dar un paso más: necesitas crear una función que **cuente cuántas canciones se reprodujeron en una ciudad específica durante un día determinado**, combinando ambos criterios de filtrado.

La función se llamará `number_tracks()` y debe aceptar dos parámetros:

- `day`: un día de la semana (por ejemplo, `'Monday'`).
- `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`).

Dentro de la función, deberás aplicar un **filtrado secuencial mediante indexación lógica**: primero tendrás que filtrar el DataFrame por el día, y luego por la ciudad. Una vez que tengas el subconjunto de datos correcto, debes contar cuántas veces aparece un valor en la columna `'user_id'`. Ese número representará el total de canciones reproducidas bajo esos dos criterios.



Etapa 3.5.
1. Declara una función llamada `number_tracks()` con dos parámetros: `day` y `city`.

2. Filtra el DataFrame para conservar solo las filas donde la columna `'day'` sea igual al valor del parámetro `day`.

3. A partir del resultado anterior, filtra nuevamente para conservar solo las filas donde la columna `'city'` sea igual al valor del parámetro `city`.

4. Extrae la columna `'user_id'` del DataFrame filtrado y utiliza el método `.count()` para contar el número de entradas.

5. Guarda ese valor en una variable y **devuélvelo** como resultado de la función.


In [22]:
df_filtrado = df[df["day"] == "Wednesday"]
df_filtrado= df_filtrado[df_filtrado["city"] == "Springfield"]
df_filtrado["user_id"].count()



np.int64(11056)

In [23]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day, city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    df_filtrado = df[df["day"] == day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    df_filtrado = df_filtrado[df_filtrado["city"] == city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    number_per_userid = df_filtrado["user_id"].count()
    # Devuelve el número de valores de la columna 'user_id'
    return number_per_userid

Etapa 3.6. Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que puedas recuperar los datos de ambas ciudades para cada uno de los tres días.

In [24]:
# El número de canciones reproducidas en Springfield el lunes
number_tracks("Monday", "Springfield")

np.int64(15740)

In [25]:
# El número de canciones reproducidas en Shelbyville el lunes
number_tracks("Monday", "Shelbyville")

np.int64(5614)

In [26]:
# El número de canciones reproducidas en Springfield el miércoles
number_tracks("Wednesday", "Springfield")

np.int64(11056)

In [27]:
# El número de canciones reproducidas en Shelbyville el miércoles
number_tracks("Wednesday", "Shelbyville")

np.int64(7003)

In [28]:
# El número de canciones reproducidas en Springfield el viernes
number_tracks("Friday", "Springfield")

np.int64(15945)

In [29]:
# El número de canciones reproducidas en Shelbyville el viernes
number_tracks("Friday", "Shelbyville")

np.int64(5895)

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

##Escribe tus conclusiones finales sobre la hipótesis.
 * ¿Los datos apoyan la idea de que el comportamiento de los usuarios respecto a la música que escuchan varía según la ciudad y el día de la semana?
 * Indica si la hipótesis debe aceptarse o rechazarse, y justifica tu respuesta con base en los resultados obtenidos.

Etapa 4. Detalla aquí tus conclusiones.

- Los datos analizados apoyan la idea de que el comportamiento musical de los usuarios no es uniforme, sino que varía según la ciudad y el día de la semana. Springfield, por ejemplo, muestra una cantidad de reproducciones significativamente mayor que Shelbyville, lo cual puede deberse a diferencias en tamaño poblacional, perfil demográfico o acceso a tecnología. Además, al observar los días de la semana, se evidencia que viernes y lunes tienen mayor actividad musical en comparación con el miércoles, lo que sugiere una influencia de los hábitos semanales y el estado de ánimo asociado a cada jornada. Estas variaciones indican que tanto el contexto geográfico como el calendario influyen directamente en el consumo de música en la plataforma.

- La hipótesis debe aceptarse, ya que los datos muestran diferencias claras en el comportamiento de los usuarios según la ciudad y el día de la semana. En términos geográficos, Springfield registra más del doble de reproducciones que Shelbyville, lo que indica que la ubicación influye directamente en el nivel de consumo musical. Por otro lado, al analizar los días de la semana, se observan variaciones significativas, con mayor actividad los viernes y lunes, y una baja notable los miércoles. Estas diferencias respaldan la hipótesis de que el comportamiento musical varía en función tanto del lugar como del momento, probablemente por factores sociales, culturales y de rutina diaria.

### Nota
En los proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)